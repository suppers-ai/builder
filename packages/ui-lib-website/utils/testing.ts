// Testing utilities for component quality assurance

/**
 * Accessibility testing utilities
 */
export class AccessibilityTester {
  /**
   * Check if element has proper ARIA attributes
   */
  static checkAriaAttributes(element: Element): string[] {
    const issues: string[] = [];

    // Check for interactive elements without proper roles
    const interactiveElements = ["button", "a", "input", "select", "textarea"];
    if (interactiveElements.includes(element.tagName.toLowerCase())) {
      if (!element.getAttribute("role") && element.tagName.toLowerCase() === "div") {
        issues.push("Interactive div element missing role attribute");
      }
    }

    // Check for images without alt text
    if (element.tagName.toLowerCase() === "img") {
      if (!element.getAttribute("alt")) {
        issues.push("Image missing alt attribute");
      }
    }

    // Check for form inputs without labels
    if (["input", "select", "textarea"].includes(element.tagName.toLowerCase())) {
      const id = element.getAttribute("id");
      const ariaLabel = element.getAttribute("aria-label");
      const ariaLabelledBy = element.getAttribute("aria-labelledby");

      if (!ariaLabel && !ariaLabelledBy) {
        if (!id || !document.querySelector(`label[for="${id}"]`)) {
          issues.push("Form input missing accessible label");
        }
      }
    }

    // Check for proper heading hierarchy
    if (element.tagName.match(/^H[1-6]$/)) {
      const level = parseInt(element.tagName.charAt(1));
      const previousHeadings = Array.from(document.querySelectorAll("h1, h2, h3, h4, h5, h6"))
        .filter((h) =>
          document.compareDocumentPosition(element) & Node.DOCUMENT_POSITION_FOLLOWING
        );

      if (previousHeadings.length > 0) {
        const lastHeading = previousHeadings[previousHeadings.length - 1];
        const lastLevel = parseInt(lastHeading.tagName.charAt(1));

        if (level > lastLevel + 1) {
          issues.push(`Heading level ${level} skips hierarchy (previous was ${lastLevel})`);
        }
      }
    }

    return issues;
  }

  /**
   * Check color contrast ratios
   */
  static checkColorContrast(element: Element): string[] {
    const issues: string[] = [];
    const styles = globalThis.getComputedStyle(element);
    const color = styles.color;
    const backgroundColor = styles.backgroundColor;

    // Simple contrast check (would need more sophisticated algorithm in production)
    if (color && backgroundColor && color !== "transparent" && backgroundColor !== "transparent") {
      const colorLuminance = this.getLuminance(color);
      const bgLuminance = this.getLuminance(backgroundColor);
      const contrast = this.getContrastRatio(colorLuminance, bgLuminance);

      if (contrast < 4.5) {
        issues.push(`Low color contrast ratio: ${contrast.toFixed(2)} (minimum 4.5)`);
      }
    }

    return issues;
  }

  /**
   * Check keyboard navigation
   */
  static checkKeyboardNavigation(container: Element): string[] {
    const issues: string[] = [];
    const focusableElements = container.querySelectorAll(
      'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])',
    );

    focusableElements.forEach((element, index) => {
      const tabIndex = element.getAttribute("tabindex");

      // Check for positive tabindex (anti-pattern)
      if (tabIndex && parseInt(tabIndex) > 0) {
        issues.push(
          `Element has positive tabindex (${tabIndex}), which can disrupt natural tab order`,
        );
      }

      // Check for focus indicators
      const styles = globalThis.getComputedStyle(element, ":focus");
      if (!styles.outline || styles.outline === "none") {
        if (!styles.boxShadow && !styles.border) {
          issues.push("Element lacks visible focus indicator");
        }
      }
    });

    return issues;
  }

  private static getLuminance(color: string): number {
    // Simplified luminance calculation
    const rgb = color.match(/\d+/g);
    if (!rgb) return 0;

    const [r, g, b] = rgb.map((c) => {
      const channel = parseInt(c) / 255;
      return channel <= 0.03928 ? channel / 12.92 : Math.pow((channel + 0.055) / 1.055, 2.4);
    });

    return 0.2126 * r + 0.7152 * g + 0.0722 * b;
  }

  private static getContrastRatio(lum1: number, lum2: number): number {
    const lighter = Math.max(lum1, lum2);
    const darker = Math.min(lum1, lum2);
    return (lighter + 0.05) / (darker + 0.05);
  }
}

/**
 * Component testing utilities
 */
export class ComponentTester {
  /**
   * Test component rendering
   */
  static testRender(component: any, props: any = {}): boolean {
    try {
      // In a real implementation, this would use a testing framework
      // For now, just check if component can be instantiated
      const instance = component(props);
      return instance !== null && instance !== undefined;
    } catch (error) {
      console.error("Component render test failed:", error);
      return false;
    }
  }

  /**
   * Test component props
   */
  static testProps(
    component: any,
    requiredProps: string[],
    optionalProps: string[] = [],
  ): string[] {
    const issues: string[] = [];

    // Test with missing required props
    try {
      component({});
      if (requiredProps.length > 0) {
        issues.push("Component should require props but accepts empty object");
      }
    } catch (error) {
      // Expected for required props
    }

    // Test with valid props
    const validProps: any = {};
    requiredProps.forEach((prop) => {
      validProps[prop] = "test-value";
    });

    try {
      const result = component(validProps);
      if (!result) {
        issues.push("Component fails to render with valid props");
      }
    } catch (error) {
      issues.push(`Component throws error with valid props: ${error.message}`);
    }

    return issues;
  }

  /**
   * Test component accessibility
   */
  static testAccessibility(element: Element): string[] {
    const issues: string[] = [];

    issues.push(...AccessibilityTester.checkAriaAttributes(element));
    issues.push(...AccessibilityTester.checkColorContrast(element));
    issues.push(...AccessibilityTester.checkKeyboardNavigation(element));

    return issues;
  }
}

/**
 * Performance testing utilities
 */
export class PerformanceTester {
  /**
   * Measure component render time
   */
  static measureRenderTime(componentFn: () => void): number {
    const start = performance.now();
    componentFn();
    const end = performance.now();
    return end - start;
  }

  /**
   * Test memory usage (simplified)
   */
  static testMemoryUsage(testFn: () => void): any {
    if ("memory" in performance) {
      const beforeMemory = (performance as any).memory.usedJSHeapSize;
      testFn();
      const afterMemory = (performance as any).memory.usedJSHeapSize;

      return {
        before: beforeMemory,
        after: afterMemory,
        difference: afterMemory - beforeMemory,
      };
    }

    return { message: "Memory measurement not available" };
  }

  /**
   * Test for memory leaks in component cleanup
   */
  static testMemoryLeaks(
    createComponent: () => any,
    destroyComponent: (component: any) => void,
  ): boolean {
    const iterations = 100;
    const components: any[] = [];

    // Create many components
    for (let i = 0; i < iterations; i++) {
      components.push(createComponent());
    }

    // Destroy all components
    components.forEach((component) => destroyComponent(component));

    // Force garbage collection if available
    if ("gc" in window) {
      (window as any).gc();
    }

    // In a real test, you'd measure memory usage here
    return true; // Simplified
  }
}

/**
 * Cross-browser compatibility testing
 */
export class BrowserCompatibilityTester {
  /**
   * Check browser feature support
   */
  static checkFeatureSupport(): Record<string, boolean> {
    return {
      cssGrid: CSS.supports("display", "grid"),
      cssFlexbox: CSS.supports("display", "flex"),
      cssCustomProperties: CSS.supports("--test", "value"),
      intersectionObserver: "IntersectionObserver" in window,
      mutationObserver: "MutationObserver" in window,
      promiseSupport: "Promise" in window,
      fetchAPI: "fetch" in window,
      webComponents: "customElements" in window,
      serviceWorker: "serviceWorker" in navigator,
      localStorage: (() => {
        try {
          localStorage.setItem("test", "test");
          localStorage.removeItem("test");
          return true;
        } catch {
          return false;
        }
      })(),
    };
  }

  /**
   * Get browser information
   */
  static getBrowserInfo(): Record<string, string> {
    const ua = navigator.userAgent;

    return {
      userAgent: ua,
      platform: navigator.platform,
      language: navigator.language,
      cookieEnabled: navigator.cookieEnabled.toString(),
      onLine: navigator.onLine.toString(),
      vendor: navigator.vendor || "unknown",
      browserName: this.detectBrowser(ua),
      browserVersion: this.detectBrowserVersion(ua),
    };
  }

  private static detectBrowser(ua: string): string {
    if (ua.includes("Chrome")) return "Chrome";
    if (ua.includes("Firefox")) return "Firefox";
    if (ua.includes("Safari") && !ua.includes("Chrome")) return "Safari";
    if (ua.includes("Edge")) return "Edge";
    if (ua.includes("Opera")) return "Opera";
    return "Unknown";
  }

  private static detectBrowserVersion(ua: string): string {
    const match = ua.match(/(?:Chrome|Firefox|Safari|Edge|Opera)\/(\d+\.?\d*)/);
    return match ? match[1] : "Unknown";
  }
}

/**
 * Automated testing suite
 */
export class AutomatedTestSuite {
  private tests: Array<{
    name: string;
    test: () => Promise<boolean>;
    category: "accessibility" | "performance" | "compatibility" | "functionality";
  }> = [];

  /**
   * Add a test to the suite
   */
  addTest(
    name: string,
    test: () => Promise<boolean>,
    category: "accessibility" | "performance" | "compatibility" | "functionality",
  ) {
    this.tests.push({ name, test, category });
  }

  /**
   * Run all tests
   */
  async runTests(): Promise<{
    passed: number;
    failed: number;
    results: Array<{ name: string; passed: boolean; error?: string; category: string }>;
  }> {
    const results = [];
    let passed = 0;
    let failed = 0;

    for (const { name, test, category } of this.tests) {
      try {
        const result = await test();
        if (result) {
          passed++;
          results.push({ name, passed: true, category });
        } else {
          failed++;
          results.push({ name, passed: false, category });
        }
      } catch (error) {
        failed++;
        results.push({
          name,
          passed: false,
          error: error instanceof Error ? error.message : "Unknown error",
          category,
        });
      }
    }

    return { passed, failed, results };
  }

  /**
   * Generate test report
   */
  generateReport(testResults: any): string {
    const { passed, failed, results } = testResults;
    const total = passed + failed;
    const passRate = ((passed / total) * 100).toFixed(1);

    let report = `# Component Library Test Report\n\n`;
    report += `**Total Tests:** ${total}\n`;
    report += `**Passed:** ${passed}\n`;
    report += `**Failed:** ${failed}\n`;
    report += `**Pass Rate:** ${passRate}%\n\n`;

    // Group by category
    const categories = ["accessibility", "performance", "compatibility", "functionality"];

    categories.forEach((category) => {
      const categoryTests = results.filter((r: any) => r.category === category);
      if (categoryTests.length === 0) return;

      report += `## ${category.charAt(0).toUpperCase() + category.slice(1)} Tests\n\n`;

      categoryTests.forEach((result: any) => {
        const status = result.passed ? "✅" : "❌";
        report += `${status} **${result.name}**\n`;
        if (result.error) {
          report += `   Error: ${result.error}\n`;
        }
        report += "\n";
      });
    });

    return report;
  }
}
