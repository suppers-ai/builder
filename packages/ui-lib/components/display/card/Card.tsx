import { ComponentChildren } from "preact";
import { BaseComponentProps } from "../../types.ts";

export interface CardProps extends BaseComponentProps {
  title?: string;
  image?: string;
  imageAlt?: string;
  compact?: boolean;
  side?: boolean;
  glass?: boolean;
  bordered?: boolean;
  actions?: ComponentChildren;
}

export function Card({
  children,
  class: className = "",
  title,
  image,
  imageAlt = "",
  compact = false,
  side = false,
  glass = false,
  bordered = false,
  actions,
  id,
  ...props
}: CardProps) {
  const cardClasses = [
    "card",
    compact ? "card-compact" : "",
    side ? "card-side" : "",
    glass ? "glass" : "",
    bordered ? "card-bordered" : "",
    "bg-base-100",
    // Enhanced shadow for DaisyUI 5 - use shadow-xl as default, but allow override
    className.includes("shadow-") ? "" : "shadow-xl",
    className,
  ].filter(Boolean).join(" ");

  return (
    <div class={cardClasses} id={id} {...props}>
      {image && (
        <figure>
          <img src={image} alt={imageAlt} />
        </figure>
      )}

      <div class="card-body">
        {title && <h2 class="card-title">{title}</h2>}
        {children}

        {actions && (
          <div class="card-actions justify-end">
            {actions}
          </div>
        )}
      </div>
    </div>
  );
}
