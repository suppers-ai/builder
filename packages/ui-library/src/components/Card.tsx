// Fresh 2.0 island component
import { useSignal } from "@preact/signals";
import { ComponentChildren } from '@json-app-compiler/shared';
import { ComponentEventHandlers, ThemedComponentProps } from '../types.ts';

export interface CardProps extends ThemedComponentProps, ComponentEventHandlers {
  variant?: 'default' | 'outlined' | 'elevated' | 'filled';
  padding?: 'none' | 'xs' | 'sm' | 'md' | 'lg' | 'xl';
  shadow?: 'none' | 'sm' | 'md' | 'lg' | 'xl';
  rounded?: 'none' | 'sm' | 'md' | 'lg' | 'xl' | 'full';
  hoverable?: boolean;
  clickable?: boolean;
  children: ComponentChildren;
  header?: ComponentChildren;
  footer?: ComponentChildren;
  title?: string;
  subtitle?: string;
  image?: string;
  imageAlt?: string;
  actions?: ComponentChildren;
}

export default function Card({
  variant = 'default',
  padding = 'md',
  shadow = 'md',
  rounded = 'md',
  hoverable = false,
  clickable = false,
  className = '',
  style = {},
  children,
  header,
  footer,
  title,
  subtitle,
  image,
  imageAlt,
  actions,
  onClick,
  onFocus,
  onBlur,
  onKeyDown,
  onKeyUp,
  onMouseEnter,
  onMouseLeave,
  id,
  'data-testid': testId,
  ...props
}: CardProps) {
  // Fresh 2.0 island state management
  const isHovered = useSignal(false);
  const isFocused = useSignal(false);
  const isPressed = useSignal(false);
  const baseClasses = [
    'bg-white',
    'transition-all',
    'duration-200',
  ];

  const variantClasses = {
    default: ['border', 'border-gray-200'],
    outlined: ['border-2', 'border-gray-300'],
    elevated: ['border-0'],
    filled: ['bg-gray-50', 'border', 'border-gray-200'],
  };

  const paddingClasses = {
    none: [],
    xs: ['p-2'],
    sm: ['p-3'],
    md: ['p-4'],
    lg: ['p-6'],
    xl: ['p-8'],
  };

  const shadowClasses = {
    none: [],
    sm: ['shadow-sm'],
    md: ['shadow-md'],
    lg: ['shadow-lg'],
    xl: ['shadow-xl'],
  };

  const roundedClasses = {
    none: [],
    sm: ['rounded-sm'],
    md: ['rounded-md'],
    lg: ['rounded-lg'],
    xl: ['rounded-xl'],
    full: ['rounded-full'],
  };

  // Fresh 2.0 island event handlers with state management
  const handleClick = (event: Event) => {
    if (clickable && onClick) {
      isPressed.value = true;
      setTimeout(() => isPressed.value = false, 150);
      onClick(event);
    }
  };

  const handleKeyDown = (event: KeyboardEvent) => {
    if (clickable && (event.key === 'Enter' || event.key === ' ')) {
      event.preventDefault();
      isPressed.value = true;
      setTimeout(() => isPressed.value = false, 150);
      onClick?.(event as any);
    }
    onKeyDown?.(event);
  };

  const handleFocus = (event: FocusEvent) => {
    isFocused.value = true;
    onFocus?.(event);
  };

  const handleBlur = (event: FocusEvent) => {
    isFocused.value = false;
    onBlur?.(event);
  };

  const handleMouseEnter = (event: MouseEvent) => {
    isHovered.value = true;
    onMouseEnter?.(event);
  };

  const handleMouseLeave = (event: MouseEvent) => {
    isHovered.value = false;
    onMouseLeave?.(event);
  };

  const interactiveClasses = [
    hoverable && isHovered.value ? 'shadow-lg' : '',
    clickable ? 'cursor-pointer focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2' : '',
    clickable && isHovered.value ? 'shadow-lg' : '',
    isPressed.value ? 'transform scale-98' : '',
    isFocused.value ? 'ring-2 ring-offset-2' : '',
  ];

  const dynamicClasses = [
    ...baseClasses,
    ...variantClasses[variant],
    ...shadowClasses[shadow],
    ...roundedClasses[rounded],
    ...interactiveClasses,
    className,
  ].filter(Boolean).join(' ');

  const contentPaddingClasses = paddingClasses[padding].join(' ');

  const CardElement = clickable ? 'button' : 'div';

  // Filter out custom props that shouldn't be passed to DOM elements
  const domProps = { ...props } as any;
  delete domProps.variant;
  delete domProps.padding;
  delete domProps.shadow;
  delete domProps.rounded;
  delete domProps.hoverable;
  delete domProps.clickable;
  delete domProps.header;
  delete domProps.footer;
  delete domProps.title;
  delete domProps.subtitle;
  delete domProps.image;
  delete domProps.imageAlt;
  delete domProps.actions;
  delete domProps.theme;
  delete domProps.size;
  delete domProps.color;
  delete domProps.children;

  return (
    <CardElement
      id={id}
      className={dynamicClasses}
      style={style}
      data-testid={testId}
      onClick={handleClick}
      onFocus={handleFocus}
      onBlur={handleBlur}
      onKeyDown={handleKeyDown}
      onKeyUp={onKeyUp}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      tabIndex={clickable ? 0 : undefined}
      role={clickable ? 'button' : undefined}
      {...domProps}
    >
      {image && (
        <div className={`${rounded !== 'none' ? 'rounded-t-' + rounded : ''} overflow-hidden`}>
          <img
            src={image}
            alt={imageAlt || ''}
            className="w-full h-48 object-cover"
          />
        </div>
      )}

      {header && (
        <div className={`${contentPaddingClasses} ${image ? 'pt-4' : ''} border-b border-gray-200`}>
          {header}
        </div>
      )}

      {(title || subtitle) && (
        <div className={`${contentPaddingClasses} ${header || image ? 'pt-4' : ''}`}>
          {title && (
            <h3 className="text-lg font-semibold text-gray-900 mb-1">
              {title}
            </h3>
          )}
          {subtitle && (
            <p className="text-sm text-gray-600">
              {subtitle}
            </p>
          )}
        </div>
      )}

      <div className={`${contentPaddingClasses} ${(title || subtitle || header || image) ? 'pt-4' : ''}`}>
        {children}
      </div>

      {actions && (
        <div className={`${contentPaddingClasses} pt-4 border-t border-gray-200 flex justify-end space-x-2`}>
          {actions}
        </div>
      )}

      {footer && (
        <div className={`${contentPaddingClasses} pt-4 border-t border-gray-200`}>
          {footer}
        </div>
      )}
    </CardElement>
  );
}