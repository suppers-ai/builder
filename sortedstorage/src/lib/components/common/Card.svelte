<script lang="ts">
	export let title = '';
	export let subtitle = '';
	export let glass = true;
	export let padding = true;
</script>

<div class="rounded-lg {glass ? 'glass' : 'bg-white dark:bg-gray-800'} {padding ? 'p-6' : ''}">
	{#if title}
		<div class="mb-4">
			<h3 class="text-lg font-semibold text-gray-900 dark:text-white">{title}</h3>
			{#if subtitle}
				<p class="text-sm text-gray-500 dark:text-gray-400">{subtitle}</p>
			{/if}
		</div>
	{/if}
	
	<div class="card-content">
		<slot />
	</div>
	
	{#if $$slots.footer}
		<div class="mt-4 pt-4 border-t border-gray-200 dark:border-gray-700">
			<slot name="footer" />
		</div>
	{/if}
</div>