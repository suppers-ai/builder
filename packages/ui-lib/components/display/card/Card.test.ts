import { assertEquals, assertStringIncludes } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { renderToString } from "preact-render-to-string";
import { h } from "preact";
import { Card } from "./Card.tsx";

Deno.test("Card - basic rendering", () => {
  const html = renderToString(h(Card, {
    children: "Card content",
  }));
  assertStringIncludes(html, "card");
  assertStringIncludes(html, "card-body");
  assertStringIncludes(html, "bg-base-100 shadow-xl");
  assertStringIncludes(html, "Card content");
});

Deno.test("Card - with title", () => {
  const html = renderToString(h(Card, {
    title: "Card Title",
    children: "Card content",
  }));
  assertStringIncludes(html, "card-title");
  assertStringIncludes(html, "Card Title");
});

Deno.test("Card - without title", () => {
  const html = renderToString(h(Card, {
    children: "Card content",
  }));
  assertEquals(html.includes("card-title"), false);
});

Deno.test("Card - with image", () => {
  const html = renderToString(h(Card, {
    image: "https://example.com/image.jpg",
    imageAlt: "Test image",
    children: "Card content",
  }));
  assertStringIncludes(html, "figure");
  assertStringIncludes(html, "https://example.com/image.jpg");
  assertStringIncludes(html, "Test image");
});

Deno.test("Card - without image", () => {
  const html = renderToString(h(Card, {
    children: "Card content",
  }));
  assertEquals(html.includes("figure"), false);
});

Deno.test("Card - image with default alt", () => {
  const html = renderToString(h(Card, {
    image: "test.jpg",
    children: "Card content",
  }));
  assertStringIncludes(html, "test.jpg");
  assertStringIncludes(html, "alt");
});

Deno.test("Card - compact variant", () => {
  const html = renderToString(h(Card, {
    compact: true,
    children: "Compact card",
  }));
  assertStringIncludes(html, "card-compact");
});

Deno.test("Card - side variant", () => {
  const html = renderToString(h(Card, {
    side: true,
    children: "Side card",
  }));
  assertStringIncludes(html, "card-side");
});

Deno.test("Card - glass variant", () => {
  const html = renderToString(h(Card, {
    glass: true,
    children: "Glass card",
  }));
  assertStringIncludes(html, "glass");
});

Deno.test("Card - bordered variant", () => {
  const html = renderToString(h(Card, {
    bordered: true,
    children: "Bordered card",
  }));
  assertStringIncludes(html, "card-bordered");
});

Deno.test("Card - with actions", () => {
  const html = renderToString(h(Card, {
    children: "Card content",
    actions: h("div", {}, [
      h("button", { class: "btn btn-primary" }, "Action"),
    ]),
  }));
  assertStringIncludes(html, "card-actions justify-end");
  assertStringIncludes(html, "btn btn-primary");
  assertStringIncludes(html, "Action");
});

Deno.test("Card - without actions", () => {
  const html = renderToString(h(Card, {
    children: "Card content",
  }));
  assertEquals(html.includes("card-actions"), false);
});

Deno.test("Card - custom class", () => {
  const html = renderToString(h(Card, {
    class: "custom-card-class",
    children: "Card content",
  }));
  assertStringIncludes(html, "custom-card-class");
});

Deno.test("Card - with id", () => {
  const html = renderToString(h(Card, {
    id: "test-card",
    children: "Card content",
  }));
  assertStringIncludes(html, 'id="test-card"');
});

Deno.test("Card - all props combined", () => {
  const html = renderToString(h(Card, {
    title: "Full Card",
    image: "image.jpg",
    imageAlt: "Card image",
    compact: true,
    side: true,
    glass: true,
    bordered: true,
    actions: h("button", { class: "btn" }, "Action"),
    class: "test-class",
    id: "full-card",
    children: "Full card content",
  }));
  assertStringIncludes(html, "card-title");
  assertStringIncludes(html, "Full Card");
  assertStringIncludes(html, "image.jpg");
  assertStringIncludes(html, "Card image");
  assertStringIncludes(html, "card-compact");
  assertStringIncludes(html, "card-side");
  assertStringIncludes(html, "glass");
  assertStringIncludes(html, "card-bordered");
  assertStringIncludes(html, "card-actions");
  assertStringIncludes(html, "test-class");
  assertStringIncludes(html, 'id="full-card"');
  assertStringIncludes(html, "Full card content");
});

Deno.test("Card - complex content", () => {
  const html = renderToString(Card({
    title: "Complex Card",
    children: h("div", {}, [
      h("p", {}, "Card description goes here."),
      h("ul", {}, [
        h("li", {}, "Feature 1"),
        h("li", {}, "Feature 2"),
        h("li", {}, "Feature 3"),
      ]),
    ]),
    actions: h("div", {}, [
      h("button", { class: "btn btn-ghost" }, "Cancel"),
      h("button", { class: "btn btn-primary" }, "Save"),
    ]),
  }));
  assertStringIncludes(html, "Complex Card");
  assertStringIncludes(html, "Card description goes here.");
  assertStringIncludes(html, "Feature 1");
  assertStringIncludes(html, "Feature 2");
  assertStringIncludes(html, "Feature 3");
  assertStringIncludes(html, "btn-ghost");
  assertStringIncludes(html, "btn-primary");
  assertStringIncludes(html, "Cancel");
  assertStringIncludes(html, "Save");
});

Deno.test("Card - image with title and content", () => {
  const html = renderToString(Card({
    title: "Image Card",
    image: "featured.jpg",
    imageAlt: "Featured image",
    children: "This card has an image, title, and content.",
  }));
  assertStringIncludes(html, "figure");
  assertStringIncludes(html, "featured.jpg");
  assertStringIncludes(html, "Featured image");
  assertStringIncludes(html, "card-title");
  assertStringIncludes(html, "Image Card");
  assertStringIncludes(html, "This card has an image, title, and content.");
});

Deno.test("Card - side card with image", () => {
  const html = renderToString(Card({
    side: true,
    image: "side-image.jpg",
    title: "Side Card",
    children: "Side layout card content",
  }));
  assertStringIncludes(html, "card-side");
  assertStringIncludes(html, "side-image.jpg");
  assertStringIncludes(html, "Side Card");
  assertStringIncludes(html, "Side layout card content");
});

Deno.test("Card - multiple action buttons", () => {
  const html = renderToString(Card({
    title: "Actions Card",
    children: "Card with multiple actions",
    actions: h("div", {}, [
      h("button", { class: "btn btn-sm btn-outline" }, "Edit"),
      h("button", { class: "btn btn-sm btn-error" }, "Delete"),
      h("button", { class: "btn btn-sm btn-primary" }, "View"),
    ]),
  }));
  assertStringIncludes(html, "card-actions justify-end");
  assertStringIncludes(html, "btn-outline");
  assertStringIncludes(html, "btn-error");
  assertStringIncludes(html, "btn-primary");
  assertStringIncludes(html, "Edit");
  assertStringIncludes(html, "Delete");
  assertStringIncludes(html, "View");
});

Deno.test("Card - glass with all features", () => {
  const html = renderToString(Card({
    glass: true,
    bordered: true,
    title: "Glass Card",
    image: "glass-bg.jpg",
    children: "Glass effect card",
    actions: h("button", { class: "btn btn-ghost" }, "Transparent"),
  }));
  assertStringIncludes(html, "glass");
  assertStringIncludes(html, "card-bordered");
  assertStringIncludes(html, "Glass Card");
  assertStringIncludes(html, "glass-bg.jpg");
  assertStringIncludes(html, "Glass effect card");
  assertStringIncludes(html, "btn-ghost");
  assertStringIncludes(html, "Transparent");
});

// Snapshot tests
Deno.test("Card - HTML snapshot simple", async (t) => {
  const html = renderToString(Card({
    title: "Simple Card",
    children: "Simple card content",
  }));
  await assertSnapshot(t, html);
});

Deno.test("Card - HTML snapshot with image", async (t) => {
  const html = renderToString(Card({
    title: "Image Card",
    image: "https://example.com/card-image.jpg",
    imageAlt: "Example image",
    children: "Card with image content",
  }));
  await assertSnapshot(t, html);
});

Deno.test("Card - HTML snapshot with actions", async (t) => {
  const html = renderToString(Card({
    title: "Action Card",
    children: "Card with actions",
    actions: h("div", {}, [
      h("button", { class: "btn btn-primary" }, "Primary"),
      h("button", { class: "btn btn-secondary" }, "Secondary"),
    ]),
  }));
  await assertSnapshot(t, html);
});

Deno.test("Card - HTML snapshot complex", async (t) => {
  const html = renderToString(Card({
    title: "Complex Card",
    image: "complex-image.jpg",
    imageAlt: "Complex card image",
    compact: true,
    bordered: true,
    children: h("div", {}, [
      h("p", { class: "text-base-content" }, "Complex card with multiple features."),
      h("div", { class: "space-y-2" }, [
        h("div", { class: "badge badge-primary" }, "Feature 1"),
        h("div", { class: "badge badge-secondary" }, "Feature 2"),
      ]),
    ]),
    actions: h("div", {}, [
      h("button", { class: "btn btn-ghost" }, "Cancel"),
      h("button", { class: "btn btn-primary" }, "Confirm"),
    ]),
    class: "custom-card",
  }));
  await assertSnapshot(t, html);
});
