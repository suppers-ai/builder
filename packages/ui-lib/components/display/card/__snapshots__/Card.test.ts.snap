export const snapshot = {};

snapshot[`CardDisplay - HTML snapshot simple 1`] = `'<div class="card bg-base-100 shadow-xl"><div class="card-body"><h2 class="card-title">Simple Card</h2>Simple card content</div></div>'`;

snapshot[`CardDisplay - HTML snapshot with image 1`] = `'<div class="card bg-base-100 shadow-xl"><figure><img src="https://example.com/card-image.jpg" alt="Example image"/></figure><div class="card-body"><h2 class="card-title">Image Card</h2>Card with image content</div></div>'`;

snapshot[`CardDisplay - HTML snapshot with actions 1`] = `'<div class="card bg-base-100 shadow-xl"><div class="card-body"><h2 class="card-title">Action Card</h2>Card with actions<div class="card-actions justify-end"><div><button class="btn btn-primary">Primary</button><button class="btn btn-secondary">Secondary</button></div></div></div></div>'`;

snapshot[`CardDisplay - HTML snapshot complex 1`] = `'<div class="card card-compact card-bordered bg-base-100 shadow-xl custom-card"><figure><img src="complex-image.jpg" alt="Complex card image"/></figure><div class="card-body"><h2 class="card-title">Complex Card</h2><div><p class="text-base-content">Complex card with multiple features.</p><div class="space-y-2"><div class="badge badge-primary">Feature 1</div><div class="badge badge-secondary">Feature 2</div></div></div><div class="card-actions justify-end"><div><button class="btn btn-ghost">Cancel</button><button class="btn btn-primary">Confirm</button></div></div></div></div>'`;
